io/src/main/scala/sbt/io/Path.scala

```diff
@@ -333,6 +367,11 @@ object PathFinder {
 
   def strict(files: Traversable[File]): PathFinder = apply(files)
 
+  case object **
+  import scala.language.dynamics
+  object * extends Dynamic {
```

Comment:
With or without dynamic, I don't know `*` as a value is good idea in sbt since there's a bunch of uses of that.
Given that we already have a filter that converts from `"*.scala"`, the creation of extension filter should be more explicit like `withExtension("*.scala")`. Hopefully Googling "withExtension sbt" would turn up more result than "sbt "*""

Reply:
I guess, in my head, it's so obvious what `foo / ** / *.scala` should mean that I hadn't really considered searchability. Does searchability matter if the syntax is put behind an import since the author of the build file has explicitly opted in?

The reason that I feel strongly about this is that because I use sbt like a shell, I like the syntax to match the shell as closely as possible. It has much lower cognitive overhead, especially when I want to interactively change the build, e.g.
```
sbt:akka> set Compile / unmanagedResources / inputs += baseDirectory.value / ** / *.txt
[info] Defining Compile / unmanagedResources / inputs
[info] The new value will be used by Compile / unmanagedResources, watchSources
[info] Reapplying settings...
[info] Set current project to akka (in build file:/Users/ethanatkins/work/scratch/akka/)
```
The syntax that you're proposing is good for a static build file where it's only typed once, but is a bit onerous to type in an interactive session.

Reply:
Perhaps the best compromise is to put the `*` syntax behind an import and add the explicit `withExtension` syntax at the sbt package level. To support the interactive use case, we'd make the `*` syntax available in the shell by default.

Reply:
Ok. I've thought a lot more about this and I think that the underlying issue is that `Source` and `PathFinder` describe completely different, though related things. The interface for `PathFinder` allows us to fetch paths, but tells us nothing about how it is doing the fetching or what it is fetching. Conversely, `Source` tells us what to fetch, but doesn't specify any implementation. This is why the `WatchSource` abstraction had to be added in the first place since `PathFinder` by definition cannot describe what sources we need to watch without eagerly fetching the files. It is possible to go `Source -> PathFinder`, but `PathFinder -> Source `is impossible in general.

Because the `Pathfinder` deliberately hides the implementation, it is impossible to apply certain kinds of optimizations on arbitrary PathFinders, e.g. proxying the requests through a cache. The `Source` api is much simpler and thus much more amenable to optimization. Moreover, watching files more or less requires the `Source` api, so that is permanently going to be part of the sbt api. It is simpler and less powerful that PathFinder which should make it easier for most users to reason about. Unfortunately without special syntax, the `Source` api is less natural to use than PathFinder when using the dsl.

The dsl for creating a `PathFinder` from a file is problematic because the inversion to `Source` isn't possible. The only way I can think of to fix this while preserving the PathFinder dsl is to partially break source compatibility and remove the implicit flag from the conversion from File -> PathFinder and add implicit conversions from `File -> Source` and `Source -> PathFinder`. To make this concrete, we'd have the following:
sbt 1.2.x
```scala
val pathFinder: PathFinder = baseDirectory.value
val files: Seq[File] = (pathFinder globRecursive AllPassFilter).get()
val alternativeFiles: Seq[File] = baseDirectory.value.allPaths.get()
```
sbt 1.3.x
```scala
val source: Source = baseDirectory.value
val files: Seq[File] = (source globRecursive AllPassFilter).get()
val alternativeFiles: Seq[File] = baseDirectory.value.allPaths.get()
```
My hope is that this would be possible in such a way that there are no binary incompatibilities and that the only source incompatibilities are places where people implicitly cast a `File` to a `PathFinder` directly. The idea is basically to just preserve the information that is lost when we convert File -> PathFinder so the dsl syntax does not actually need to change.

One possible alternative that I think would be less likely to cause source incompatibilities would be to deprecate the `Source` class and introduce a new class
```
  final class PathFinderInput(val base: File, val filter: FileFilter, val recursive: Boolean) extends PathFinder
```
The reason I'd like to replace `Source` is that the name is singular but it actually describes a collection of `File`s. Moreover, I find the excludeFilter parameter unnecessary and confusing. While it is technically true that there are conditions that cannot be expressed without the excludeFilter, the vast majority of the time (and everywhere that I've seen it used in sbt), the exclude case can just be handled by replacing the inputFilter with inputFilter -- excludeFilter.

I intend to deprecate `watchSources` anyway for the new watch implementation and it will be trivial to convert between `PathFinderInput` and `Source`. Also `Source` is defined in the sbt.internal.io package and I'd imagine `PathFinderInput` would probably be in the sbt or sbt.io package. The `PathFinderInput` name is not my favorite, so I'm trying to think of a better one. I considered `Inputs`, but, in practice, I expect to mostly use collections of `PathFinderInput` and I find `val inputs = settingKey[Seq[Inputs]]("...")` a bit confusing.

I will continue working on my sbt PR and will push some commits that implement this idea both with the `Source` class and with the new `PathFinderInput` class so we can see what the change would look like.

Reply:
`PathFinderInput` route sounds like a better plan, especially if the build user can opt into migrating from old `Source` style.

As per name is concerned, I naming it with the metaphor that's close to the build user's mental image would be good. How about something like `DirectoryWatch`, kind of giving hat tip to takari/directory-watcher? Or `WatchSet`.

> Also `Source` is defined in the `sbt.internal.io` package and I'd imagine `PathFinderInput` would probably be in the `sbt` or `sbt.io package`. 

Yea, it's an odd one. It's in `sbt.internal.io` but there's an alias for it called `sbt.WatchSource` - https://github.com/sbt/sbt/blob/v1.2.7/sbt/src/main/scala/sbt/Import.scala#L60-L61